More explanation about the APLN token "h3" (#375)
quicwg/ops-drafts: draft-ietf-quic-applicability.md

practice that a port number maps unambiguously to an application, the
use of ports that cannot easily be mapped to a registered service name
might lead to blocking or other changes to the forwarding behavior by network
elements such as firewalls that use the port number for application
identification.

Applications could define an alternate endpoint discovery mechanism to allow
the usage of ports other than the default. For example, HTTP/3 ({{Sections 3.2
and 3.3 of QUIC-HTTP}}) specifies the use of HTTP Alternative Services
{{?RFC7838}} for an HTTP origin to advertise the availability of an equivalent
HTTP/3 endpoint on a certain UDP port by using the "h3" Application-Layer
Protocol Negotiation (ALPN) {{?RFC7301}} token.

ALPN permits the
client and server to negotiate which of several protocols will be used on a
given connection.  Therefore, multiple applications might be supported on a
single UDP port based on the ALPN token offered.  Applications using QUIC
are required to register an ALPN token for use in the TLS handshake.

As QUIC version 1 deferred defining a complete version negotiation mechanism,
HTTP/3 requires QUIC version 1 and defines the

> ALPN token ("h3") to only apply to that version. 

Author: ```suggestion
ALPN token ("h3") to only apply to that version.
```

the default port for HTTP/3 {{QUIC-HTTP}} is UDP port 443, analogous to HTTP/1.1
or HTTP/2 over TLS over TCP.

Given the prevalence of the assumption in network management
practice that a port number maps unambiguously to an application, the
use of ports that cannot easily be mapped to a registered service name
might lead to blocking or other changes to the forwarding behavior by network
elements such as firewalls that use the port number for application
identification.

Applications could define an alternate endpoint discovery mechanism to allow
the usage of ports other than the default. For example, HTTP/3 ({{Sections 3.2
and 3.3 of QUIC-HTTP}}) specifies the use of HTTP Alternative Services
{{?RFC7838}} for an HTTP origin to advertise the availability of an equivalent
HTTP/3 endpoint on a certain UDP port by using the "h3" Application-Layer
Protocol Negotiation (ALPN) {{?RFC7301}} token.

> Additionally, ALPN permits the

Reviewer: ```suggestion
ALPN permits the
```

client and server to negotiate which of several protocols will be used on a
given connection.  Therefore, multiple applications might be supported on a
single UDP port based on the ALPN token offered.  Applications using QUIC

> should register an ALPN token for use in the TLS handshake.

Reviewer: ```suggestion
are required to register an ALPN token for use in the TLS handshake.
```

ALPN - or something like it - is mandatory in QUIC.  But when TLS is used - and this is about QUIC v1 where TLS is used - that effectively means ALPN is mandatory.

As QUIC version 1 deferred defining a complete version negotiation mechanism,
HTTP/3 requires QUIC version 1 and defines the

> ALPN token ("h3") to only apply to that version. This approach avoids
> ambiguity of the used version in absence of a version negotiation
> mechanism. So far no single approach has been selected for

Reviewer: ```suggestion
ALPN token ("h3") to only apply to that version. 
So far no single approach has been selected for
```
The ambiguity exists as long as there is even the potential for another QUIC version.  Safer not to say anything.

managing the use of different QUIC versions, neither in HTTP/3 nor in general.

> Coupling QUIC version to ALPN tokens is one approach but not the
> only approach. Other options may be suitable when a version negotiation
> scheme gets developed and new protocol versions (at both application and
> transport layers) are defined. Application protocols that use QUIC need to

Reviewer: ```suggestion
Application protocols that use QUIC need to
```

I don't see there being any value in talking about this coupling further.

consider how the protocol will manage different QUIC versions.
Decisions for those protocols might be informed by choices made by other

> protocols, like HTTP/3.

Reviewer 2: I propose that you simply delete this paragraph.

First, I disagree that there is "ambiguity of the used version." The version field is quite explicit; the only thing QUIC is missing is downgrade detection.

More broadly, I don't think we have a consensus on how ALPN will work with QUIC versions, so it is not possible to say anything useful here, unless we're going to have that conversation before this ships.

The rest of the new text is great.

Author: This is all of the new text (I just move the other paragraphs around to make things fit). If we don't want this new part, I would like to keep this sentence from the old text instead because I think that's the important bit of information:

`Note that HTTP/3's ALPN token ("h3") identifies not only the version of the application protocol, but also the version of QUIC itself`

We can remove the part about ambiguity...

Reviewer 2: Well this is the part that I disagree with. I would like QUICv2 to also use the h3 token, and don't see why that's incompatible with the text in base-drafts.

Reviewer: I think that there is value in the statement; that it is currently true is useful information, but the text doesn't say that it is necessarily going to stay that way.

If the objection is to the implication, would it be more acceptable if it said "HTTP/3 requires QUIC version 1 and defines the ALPN token ("h3") to only apply to that version.  This decision could be revised when new QUIC versions are defined."

Reviewer 2: While I think we have a lot of basic decisions still to make about ALPN and versioning, the warning at the end of your proposed text does make it clear that this is all in flux. I can live with it.

the usage of ports other than the default. For example, HTTP/3 ({{Sections 3.2
and 3.3 of QUIC-HTTP}}) specifies the use of HTTP Alternative Services
{{?RFC7838}} for an HTTP origin to advertise the availability of an equivalent

> HTTP/3 endpoint on a certain UDP port by using the "h3" ALPN token.  

Author: ```suggestion
HTTP/3 endpoint on a certain UDP port by using the "h3" ALPN token.
```
